File: benches/comparative/typescript/bench_polymarket.mjs
// Polymarket REST API head-to-head: OpenPX vs @polymarket/clob-client.
//
// 20 iterations × 100 ms gap, polyfill methodology. Writes a JSON
// summary to stdout that the README renderer parses.
//
// Run: `node bench_polymarket.mjs > ../results/typescript_polymarket.json`

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Bench } from "tinybench";
import { Exchange } from "@openpx/sdk";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const META = JSON.parse(
  fs.readFileSync(
    path.join(__dirname, "..", "fixtures", "polymarket_book.meta.json"),
    "utf8"
  )
);
const ASSET_ID = META.asset_id;
const CONDITION_ID = META.condition_id;
const GAP_MS = 100;

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// Optional comparison target — skip the SDK column gracefully if not installed.
let ClobClient = null;
try {
  ({ ClobClient } = await import("@polymarket/clob-client"));
} catch {
  console.error("note: @polymarket/clob-client not installed, OpenPX-only run");
}

const openpx = new Exchange("polymarket", {});
const sdk = ClobClient ? new ClobClient("https://clob.polymarket.com") : null;

const bench = new Bench({ time: 0, iterations: 20 });

// --- fetch_markets ---------------------------------------------------------
bench.add("openpx::polymarket::fetch_markets", async () => {
  await sleep(GAP_MS);
  await openpx.fetchMarkets();
});
if (sdk) {
  bench.add("polymarket-clob-client::fetch_markets", async () => {
    await sleep(GAP_MS);
    await sdk.getSamplingMarkets();
  });
}

// --- fetch_market ----------------------------------------------------------
bench.add("openpx::polymarket::fetch_market", async () => {
  await sleep(GAP_MS);
  await openpx.fetchMarket(CONDITION_ID);
});
if (sdk) {
  bench.add("polymarket-clob-client::fetch_market", async () => {
    await sleep(GAP_MS);
    await sdk.getMarket(CONDITION_ID);
  });
}

// --- fetch_orderbook -------------------------------------------------------
bench.add("openpx::polymarket::fetch_orderbook", async () => {
  await sleep(GAP_MS);
  await openpx.fetchOrderbook(ASSET_ID);
});
if (sdk) {
  bench.add("polymarket-clob-client::fetch_orderbook", async () => {
    await sleep(GAP_MS);
    await sdk.getOrderBook(ASSET_ID);
  });
}

// --- fetch_trades ----------------------------------------------------------
bench.add("openpx::polymarket::fetch_trades", async () => {
  await sleep(GAP_MS);
  await openpx.fetchTrades(ASSET_ID);
});
if (sdk) {
  bench.add("polymarket-clob-client::fetch_trades", async () => {
    await sleep(GAP_MS);
    // Trade history is exposed via market trades on the SDK
    await sdk.getMarketTradesEvents(CONDITION_ID);
  });
}

await bench.run();

const tasks = bench.tasks.map((t) => ({
  name: t.name,
  mean_ns: t.result?.mean ? t.result.mean * 1e6 : null,
  stddev_ns: t.result?.sd ? t.result.sd * 1e6 : null,
  samples: t.result?.samples?.length ?? 0,
}));

console.log(JSON.stringify({ tasks }, null, 2));

File: benches/comparative/python/bench_polymarket.py
"""Polymarket REST head-to-head: OpenPX vs py-clob-client.

Single fair comparison: `fetch_orderbook`. Both libraries hit the same
upstream endpoint (`/book?token_id=...`) and return the same shape, so
the ratio reflects real client-side overhead — not endpoint choice.

`fetch_markets` was deliberately dropped: OpenPX hits Gamma
`/events/keyset` (rich event tree → typed unified Market) while
py-clob hits CLOB `/sampling-markets` (raw dicts, 60% smaller payload).
Different operations, can't fairly compare.

20 iterations × 100 ms gap, polyfill methodology.

Run: `pytest benches/comparative/python/bench_polymarket.py --benchmark-only`
"""

from __future__ import annotations

import json
import time
from pathlib import Path

import pytest

from openpx import Exchange

try:
    from py_clob_client.client import ClobClient
    from py_clob_client.constants import POLYGON

    HAS_PYCLOB = True
except ImportError:  # pragma: no cover - optional
    HAS_PYCLOB = False


FIXTURE = Path(__file__).resolve().parents[1] / "fixtures" / "polymarket_book.meta.json"
HOST = "https://clob.polymarket.com"
GAP = 0.1


def _asset_id() -> str:
    return json.loads(FIXTURE.read_text())["asset_id"]


@pytest.fixture(scope="module")
def openpx_exchange() -> Exchange:
    return Exchange("polymarket")


@pytest.fixture(scope="module")
def pyclob_client() -> "ClobClient":
    if not HAS_PYCLOB:
        pytest.skip("py-clob-client not installed")
    return ClobClient(HOST, chain_id=POLYGON)


def test_openpx_fetch_orderbook(benchmark, openpx_exchange):
    aid = _asset_id()

    def run():
        time.sleep(GAP)
        return openpx_exchange.fetch_orderbook(aid)

    benchmark.pedantic(run, iterations=1, rounds=20)


def test_pyclob_fetch_orderbook(benchmark, pyclob_client):
    aid = _asset_id()

    def run():
        time.sleep(GAP)
        return pyclob_client.get_order_book(aid)

    benchmark.pedantic(run, iterations=1, rounds=20)
